R/helper_database_functions.R

```diff
@@ -116,23 +114,41 @@ update_db <- function(dbdir = ".",
 }
 
 # this is a helper function to build nflfastR database from Scratch
-build_db <- function(tblname = "nflfastR_pbp", db_conn) {
-  if (DBI::dbExistsTable(db_conn, tblname)) {
-    message(glue::glue("Purging old {tblname} table from database..."))
-    DBI::dbRemoveTable(db_conn, tblname)
-  }
+build_db <- function(tblname = "nflfastR_pbp", db_conn, rebuild = FALSE) {
 
-  games <- readRDS(url("https://github.com/leesharpe/nfldata/blob/master/data/games.rds?raw=true"))
-  complete_games <- games %>%
+  valid_seasons <- readRDS(url("https://github.com/leesharpe/nfldata/blob/master/data/games.rds?raw=true")) %>%
     dplyr::filter(.data$season >= 1999 & !is.na(.data$result)) %>%
-    dplyr::arrange(.data$gameday)
-
-  seasons <- 1999:dplyr::last(complete_games$season)
-  message(glue::glue("Starting download of {length(seasons)} seasons between {min(seasons)} and {max(seasons)}..."))
-  progressr::with_progress({
-    p <- progressr::progressor(along = seasons)
-    purrr::walk(seasons, load_cleaned_pbp, db_conn, p, tblname)
-  })
+    dplyr::group_by(.data$season) %>%
+    dplyr::summarise() %>%
+    dplyr::ungroup()
+
+  if (all(rebuild == TRUE)) {
+    message(glue::glue("Purging all rows from {tblname} in your connected database..."))
+    DBI::dbExecute(db_conn, glue::glue("DELETE FROM {tblname}"))
+    seasons <- valid_seasons %>% dplyr::pull("season")
+    message(glue::glue("Starting download of {length(seasons)} seasons between {min(seasons)} and {max(seasons)}..."))
+  } else if (is.numeric(rebuild) & any(rebuild %in% valid_seasons$season)) {
+    string <- paste0(rebuild, collapse = ", ")
+    message(glue::glue("Purging {string} season(s) from {tblname} in your connected database..."))
+    DBI::dbExecute(db_conn, glue::glue("DELETE FROM {tblname} WHERE season IN ({string})"))
```

Comment:
this should be a glue_sql I think: 

```
DBI::dbExecute(db_conn, glue::glue_sql("DELETE FROM {DBI::SQL(tblname)} WHERE season IN({string *})", .con = db_conn))
```
which will correctly interpolate multiple seasons into the IN operator

Reply:
Oh I see you paste earlier - it's still probably safer to refer directly to the rebuild/valid_seasons etc

Reply:
May also be worth checking that the rebuild argument is logical, numeric, or "NEW" :)

Reply:
> May also be worth checking that the rebuild argument is logical, numeric, or "NEW" :) 

This should be done indirectly via the if else conditions. 

I need those conditions anyway which is the reason why I didn't add this check.

Reply:
> this should be a glue_sql I think: 
> 
> 
> 
> ```
> 
> DBI::dbExecute(db_conn, glue::glue_sql("DELETE FROM {DBI::SQL(tblname)} WHERE season IN({string *})", .con = db_conn))
> 
> ```
> 
> which will correctly interpolate multiple seasons into the IN operator, rather than using paste0 (this helps defend against SQL injections!)

I had no idea there is something like `glue_sql()`

Reply:
I foresee one problem with this change. It removes the ability to run `update_db` with `force_rebuild=TRUE` to successfully rebuild the database between breaking changes of the data - i.e. new columns added or column type changes in the data. This obviously isn't a super common occurrence but with package development being fast-moving (a good thing!) it was a nice feature to have so that you didn't need to worry about needing to drop your DB between releases.

Reply:
> I foresee one problem with this change. It removes the ability to run `update_db` with `force_rebuild=TRUE` to successfully rebuild the database between breaking changes of the data - i.e. new columns added or column type changes in the data. This obviously isn't a super common occurrence but with package development being fast-moving (a good thing!) it was a nice feature to have so that you didn't need to worry about needing to drop your DB between releases.

Yep this is a good point. For what it's worth, we're planning on freezing the columns (both number of columns and what is in them) during the season so we don't break things for people. There's a tradeoff between complexity (adding yet another option in this db function) versus flexibility that I don't really know the answer to.

Reply:
> I foresee one problem with this change. It removes the ability to run `update_db` with `force_rebuild=TRUE` to successfully rebuild the database between breaking changes of the data - i.e. new columns added or column type changes in the data. This obviously isn't a super common occurrence but with package development being fast-moving (a good thing!) it was a nice feature to have so that you didn't need to worry about needing to drop your DB between releases.  

I agree with this and think the solution is quite easy. 
We return to removing the whole table with `force_rebuild = TRUE` (to be able to add new columns or types) and keep removing rows for `force_rebuild = {some_seasons}`. 
And if someone wants to update all seasons but without killing the rows or types he/she can do `force_rebuild = 1999:2020` for example.